File: TrappingRainWater.cpp
Given n non-negative integers representing an elevation map where the width of each bar is 1, 
compute how much water it is able to trap after raining.

The above elevation map is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. 
In this case, 6 units of rain water (blue section) are being trapped. 

Example:

Input: [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6

class Solution 
{
public:
    int trap(vector<int>& height) 
    {
        int result=0;
        /*
        stack<int>s;
        for(int i=0;i<height.size();i++)
        {
            while(!s.empty() && height[i]>height[s.top()])
            {
                int top=s.top();
                s.pop();
                if(s.empty())
                    break;
                int dist=i-s.top()-1;
                int boundedHeight=min(height[i],height[s.top()])-height[top];
                result+=dist*boundedHeight;
            }
            s.push(i);
        }
        */
        int left=0,right=height.size()-1;
        int leftMax=0,rightMax=0;
        while(left<right)
        {
            if(height[left]<height[right])
            {
                if(height[left]>=leftMax)
                    leftMax=height[left];
                else
                    result+=leftMax-height[left];
                left++;
            }
            else
            {
                if(height[right]>=rightMax)
                    rightMax=height[right];
                else
                    result+=rightMax-height[right];
                right--;
            }
        }
        return result;
    }
};

File: GeeksForGeeks/TrappingRainWater.py
"""
Given an array arr[] of N non-negative integers representing the height of blocks. If width of each block is 1,
compute how much water can be trapped between the blocks during the rainy season.


Example 1:

Input:
N = 6
arr[] = {3,0,0,2,0,4}
Output:
10
Explanation:

Example 2:

Input:
N = 4
arr[] = {7,4,0,9}
Output:
10
Explanation:
Water trapped by above
block of height 4 is 3 units and above
block of height 0 is 7 units. So, the
total unit of water trapped is 10 units.
Example 3:

Input:
N = 3
arr[] = {6,9,9}
Output:
0
Explanation:
No water will be trapped.
"""


class Solution:
    def trappingWater(self, arr, n):
        # Code here
        left = [0 for i in range(n)]
        right = [0 for i in range(n)]
        water = 0
        left[0] = arr[0];
        for i in range(1, n):
            left[i] = max(left[i - 1], arr[i])
        right[n - 1] = arr[n - 1]
        for i in reversed(range(n - 1)):
            right[i] = max(right[i + 1], arr[i])
        for i in range(n):
            water += min(left[i], right[i]) - arr[i]
        return water;


# {
#  Driver Code Starts
# Initial Template for Python 3

import math


def main():
    T = int(input())
    while (T > 0):
        n = int(input())

        arr = [int(x) for x in input().strip().split()]
        obj = Solution()
        print(obj.trappingWater(arr, n))

        T -= 1


if __name__ == "__main__":
    main()

# } Driver Code Ends